joomla/joomla-cms
# [4.0] Download key manager - GSoC Expand Extensions Manager Redo

### Redo
This is a redo of #17575 as I can no longer push to the original repo.

### Introduction
Currently there is no clear overview of which extensions have download keys to make users download their extensions with an extra_query with their keys included to allow the download. Developers have been currently hacking this system by creating their own view to add a download key to the extra_query field in the #__update_sites table (Example of an extra_query field: dlid=`key`&amp;dummy=my.zip).
Each extension developer may have a different way to build this extra query so the idea is to make a view to manage and insert the download keys of all extensions in a single view in com_installer.

To set or customize the prefix and suffix of the download key you need to have an <dlid> tag with prefix and suffix in your installation file like this:

`<dlid prefix="dlid=" suffix="&amp;dummy=my.zip"/>`
(note: remember to close />)

If an extension has no download ID set in the manifest file, the download key will show up as N/A in the list and will not be editable.

### Testing Instructions

Here are the extensions to test the feature. I have not included all types as I think these 3 already make clear what this is about.
1. Run the [MySQL update query](https://github.com/joomla/joomla-cms/pull/25553/files#diff-adc3eedd33560141a69c0cb8d1d4214e) or the [Postgres update query](https://github.com/joomla/joomla-cms/pull/25553/files#diff-7b5d5176f5c5de2280a23dcd12f034ea) depending on the database you use
2. Apply the patch
3. Install the extensions below
Component
[keycomponent.zip](https://github.com/joomla/joomla-cms/files/3390080/keycomponent.zip)
Plugin:
[keyplugin.zip](https://github.com/joomla/joomla-cms/files/3390082/keyplugin.zip)
Module:
[keymodule.zip](https://github.com/joomla/joomla-cms/files/3390083/keymodule.zip)
4. Go to System -> Update Sites
<img width="619" alt="image" src="https://user-images.githubusercontent.com/359377/63717106-c67fac80-c847-11e9-8d03-7dba94317f30.png">

The list shows just the name of the extension and the update site URL.

5. Clicking on the extension name will open the Edit screen.
<img width="529" alt="image" src="https://user-images.githubusercontent.com/359377/63718231-62aab300-c84a-11e9-9299-620191b59053.png">
5. Enter a download Key
6. Click on Save & Close
7. The download key will now be shown in the list
<img width="436" alt="image" src="https://user-images.githubusercontent.com/359377/63718442-e06ebe80-c84a-11e9-8de6-a63da2646602.png">

The Download Key is shown without the prefix and suffix. Only the developer will be responsible to define the prefix and suffix of the download key.

You are going to need a real extension with a download key to see if the download key is send. For now you could also check the database by doing this:
1. Go to the table __update_sites table
2. Check the field `extra_query` for the extension where you added the download key. It should have the format of `prefix` `key` `suffix`

### Expected result

A simple a intuitive way to the user manage his download keys and less work to the developers to create a view to generate the extra_query

### Actual result
There is no single place to manage your download keys.

### Documentation Changes Required
Yes, once merged I will write a document on how to incorporate the tags in the installer XML
